### ISSUE TITLE
parameter destructuring renaming and shadowed variable

### ISSUE THREAD

**Issue by lbesson** [palantir/tslint#727]
Following code is incorrectly handled by no-shadowed-variable rule
```ts
interface MyInterface  {
  public a: number;
}

function myFunction(
  { a: var1 }: MyInterface,
  { a: var2 }: MyInterface, // shouldn't create failure
  { a: var1 }: MyInterface // should create failure
) {
  var1 = 2;
  var2 = 3;
}
```
In fact, failure seems to be created because when visiting parameter destructuring renaming, variableName is undefined (so we get shadowed variable: 'undefined'

As an aside, above code also raises no unused variable false positives

**Comment by lbesson** [comment]
In fact, it seems that in that case variables are present in node.name.elements

I can try to propose a PR for this if you're ok with it

**Comment by adidahiya** [comment]
Just to confirm, are you on the latest TSLint? [This PR](https://github.com/palantir/tslint/pull/616) fixed some related issues in `no-shadowed-variable`, but the bug you pointed out might still be an issue.

**Comment by lbesson** [comment]
On master. I can't check against next branch right now but I'll let you
know.

**Comment by adidahiya** [comment]
ok, that shouldn't be necessary; all the fixes we've done have been committed to master. I'll categorize this as a bug

**Comment by alexlrobertson** [comment]
I'm encountering the same issue.

```js
interface IThing {
    id: string;
    color: string;
}

function getFirstOfColor(color: string) {
    return _([{color: 'foo'}, {color: 'bar'}])
        .filter(({ thingColor }: IThing): boolean => thingColor === color)
        // ^ TSLint: shadowed variable: 'undefined'
        .first();
}

// tslint@2.5.1
```
